src/server/registry/tag/tag.go

```diff
@@ -15,22 +15,165 @@
+func NewHandler(repoMgr repository.Manager, tagMgr tag.Manager) http.Handler {
+	return &handler{
+		repoMgr: repoMgr,
+		tagMgr:  tagMgr,
+	}
 }
 
-type handler struct{}
+type handler struct {
+	repoMgr repository.Manager
+	tagMgr  tag.Manager
 
+	repositoryName string
+}
+
+// ServeHTTP ...
 func (h *handler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
-	// TODO implement
-	// support pagination, this is required for OCI registry spec
+	switch req.Method {
+	case http.MethodGet:
+		h.get(w, req)
+	}
+}
+
+// get return the list of tags
+
+//Content-Type: application/json
+//Link: <<url>?n=<n from the request>&last=<last tag value from previous response>>; rel="next"
+//
+//{
+//    "name": "<name>",
+//    "tags": [
+//      "<tag>",
+//      ...
+//    ]
+//}
+func (h *handler) get(w http.ResponseWriter, req *http.Request) {
+	var withPagination = true
+
+	reqQ := req.URL.Query()
+	lastEntry := reqQ.Get("last")
+	maxEntries, err := strconv.Atoi(reqQ.Get("n"))
+	if err != nil || maxEntries < 0 {
+		err := ierror.New(err).WithCode(ierror.BadRequestCode).WithMessage("the N must be a positive int type")
+		reg_error.Handle(w, req, err)
+		return
+	}
+	if maxEntries == 0 {
+		withPagination = false
+	}
+
+	var repoID int64
+	var tagNames []string
+	vars := mux.Vars(req)
+	h.repositoryName = vars["name"]
+
+	repoID, err = h.getRepoID(req)
+	if err != nil {
+		reg_error.Handle(w, req, err)
+		return
+	}
+
+	// get tags ...
+	total, tags, err := h.tagMgr.List(req.Context(), &q.Query{
+		Keywords: map[string]interface{}{
+			"repository_id": repoID,
+		},
+	})
+	if err != nil {
+		reg_error.Handle(w, req, err)
+		return
+	}
+	if total == 0 {
+		h.sendResponse(w, req, tagNames)
+		return
+	}
+
+	for _, tag := range tags {
+		tagNames = append(tagNames, tag.Name)
+	}
+	sort.Strings(tagNames)
+
+	// with "last", if last page, no link header
+	// without "last", if one page covers "n", no link header
+	if lastEntry != "" {
+		lastEntryIndex := util.IndexString(tagNames, lastEntry)
+		if len(tagNames)-(lastEntryIndex+1) < maxEntries {
+			withPagination = false
+		}
+		tagNames = tagNames[lastEntryIndex+1 : lastEntryIndex+maxEntries]
+	} else {
+		if len(tagNames) <= maxEntries {
+			withPagination = false
+		}
+		tagNames = tagNames[0:maxEntries]
+	}
+
+	if withPagination {
+		lastEntry = tagNames[len(tagNames)-1]
+		urlStr, err := util.CreateLinkEntry(req.URL.String(), maxEntries, lastEntry)
+		if err != nil {
+			reg_error.Handle(w, req, err)
+			return
+		}
+		w.Header().Set("Link", urlStr)
+	}
+
+	h.sendResponse(w, req, tagNames)
+	return
+}
+
+// getRepoID ...
+func (h *handler) getRepoID(req *http.Request) (int64, error) {
+	var repoID int64
+	total, repoRecord, err := h.repoMgr.List(req.Context(), &q.Query{
+		Keywords: map[string]interface{}{
+			"name": h.repositoryName,
+		},
+	})
+	if err != nil {
+		return 0, err
+	}
+	if total <= 0 {
+		err := ierror.New(errors.New("repositoryNotFound")).WithCode(ierror.NotFoundCode)
```

Comment:
This is where the error generated, so I think the original cause should be nil:
`ierror.New(nil).WithCode(xxx).WithMessage(xxx)`?
Specify what repository name is in the massage

Reply:
updated.